### PR TITLE
NONPR-1991 NRS retrieves and stores attachment from given location

### PR DESCRIPTION
**Pull request by greg-gajda**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1,8 +1,8 @@
 enablePlugins(GitVersioning)
 enablePlugins(BuildInfoPlugin)
 
-val akkaHttpVersion = "10.2.4"
-val akkaVersion = "2.6.14"
+val akkaHttpVersion = "10.2.7"
+val akkaVersion = "2.6.18"
```
Comment (PaulHodgson) [inline build.sbt:5]: +1

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/server/Routes.scala [hunk 2]
```diff
@@ -63,7 +63,7 @@ class Routes(flow: AttachmentFlow)(implicit val system: ActorSystem[_], config:
                       .log(name = "attachmentFlow")
                       .addAttributes(logLevels(onElement = Off, onFinish = Info, onFailure = Error))
                       .map(IncomingRequest(ApiKey(apiKey), _))
-                      .via(flow.validationFlow)
+                      .via(flow.applicationFlow)
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/server/Routes.scala:66]: 👍

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import akka.util.ByteString
+import org.apache.commons.codec.digest.DigestUtils
+import uk.gov.hmrc.nonrep.attachment.models.AttachmentRequestKey
+import uk.gov.hmrc.nonrep.attachment.server.ServiceConfig
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.util.Try
+
+trait Storage[A] {
+  def s3Call(implicit system: ActorSystem[_])
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:22]: This particular def doesn't seem to be related to s3 and would work for any request I think. Perhaps rename to something which makes it clearer what's happening here? Something like `requestToConnectionPool` maybe?

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+class StorageService extends Storage[AttachmentRequestKey] {
+  import Storage._
+
+  def s3Call(implicit system: ActorSystem[_])
+  : Flow[(HttpRequest, EitherErr[AttachmentRequestKey]), (Try[HttpResponse], EitherErr[AttachmentRequestKey]), Any] =
+    Http().superPool[EitherErr[AttachmentRequestKey]]()
+
+  def s3Request(data: EitherErr[AttachmentRequestKey]): HttpRequest =
+    data.map(attachment => Get(Uri(attachment.request.attachmentUrl))).toOption.getOrElse(HttpRequest())
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:42]: I think it would it be better just to fail fast if for some reason `toOption` returns `None`. 

Maybe something like

`Get(Uri(attachment.request.attachmentUrl))).toOption.getOrElse(throw new RuntimeException("failed to create s3 request"))`

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+  def putS3Object(attachment: AttachmentRequestKey, file: ByteString)
+                 (implicit system: ActorSystem[_], config: ServiceConfig): Future[EitherErr[AttachmentRequestKey]] = Source
+    .single(file)
+    .runWith(S3
+      .multipartUpload(config.attachmentsBucket, attachment.request.attachmentId)
+      .mapMaterializedValue(_.filter(_.eTag == file.toArray[Byte].calculateMD5)
+        .map(_ => Right(attachment).withLeft[ErrorMessage])))
+}
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:65]: I don't think we'll get a clear error message out of `putS3Object` in the strange event that the eTag and the  calculated md5 don't match.

Despite what was said earlier about testing, I think it might be worth splitting the code to compare the eTag and md5 into a distinct function which could be usefully unit tested for success and fail case. This can be usefully tested without worrying about s3 or using a mock.

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+object Storage {
+
+  implicit val defaultStorageService: StorageService = new StorageService()
+
+  implicit class ArrayOfBytesWithSha256(input: Array[Byte]) {
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:71]: this duplicates (new) code in the ApiKey object. Perhaps time to add `CryptoUtils` (and test?).

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala [hunk 7]
```diff
@@ -98,34 +156,37 @@ class AttachmentFlow()(implicit val system: ActorSystem[_], config: ServiceConfi
                                                                                                   | ~> startMeasureEsCall ~> zipEsCalls.in0 ~> zipEsCalls.out ~> Sink.ignore
 								                                                                                  | ~> parseEsResponse  ~> collectEsMetrics ~> out
     */
-  val validationFlow: Flow[IncomingRequest, EitherErr[AttachmentRequest], NotUsed] =
+  val applicationFlow: Flow[IncomingRequest, EitherErr[AttachmentRequest], NotUsed] =
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala:159]: 👍

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala [hunk 8]
```diff
@@ -98,34 +156,37 @@ class AttachmentFlow()(implicit val system: ActorSystem[_], config: ServiceConfi
...
 
         val validateXApiHeaderShape = builder.add(validateXApiHeader)
-        val validationShape = builder.add(validateRequest)
-        val responseShape = builder.add(remapAttachmentRequestKey)
+        val responseRemappedShape = builder.add(remapAttachmentRequestKey)
 
-        validateXApiHeaderShape ~> validationShape
-        validationShape ~> partitionEsCalls
-        partitionEsCalls ~> merge
+        validateXApiHeaderShape ~> validateRequest ~> partitionEsCalls
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala:176]: good catch

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 3 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:22]: I'll refactor it a bit more, there's too many similarities between index and storage interface to leave it like that

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 5 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:65]: sure, it makes sense

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 6 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:71]: :+1:

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 4 above)
Comment (PaulHodgson) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:42]: You could maybe return the `Either[ErrorMessage, HttpRequest]` but that might get noisy. The exception has the advantage of failing fast while simplifying the return type.

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 5 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:63]: it's been done in the correct order now, first MD5 calculation and then sending it with Content-MD5 header

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 3 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala]: Please take another look @PaulHodgson - there significant number of changes introduced, but there's some reusability of interfaces that serve the same purpose

Comment (PaulHodgson) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala]: I hadn't considered the overlap with `Index` when I made the original comment.

I definitely prefer the newer approach and the naming convention in particular is better now which is what this comment was about.

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala (hunk 5 above)
Comment (PaulHodgson) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Storage.scala:63]: The new approach looks better to me.

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala [hunk 9]
```diff
@@ -24,101 +24,59 @@ import uk.gov.hmrc.nonrep.attachment.utils.JsonFormats._
...
+    if (response.status == StatusCodes.OK) {
+      import system.executionContext
+      response.entity.dataBytes
+        .runFold(ByteString.empty)(_ ++ _)
+        .map(bytes => (bytes.utf8String.parseJson.convertTo[SearchResponse], bytes))
+        .map(Right(_).withLeft[ErrorMessage])
+        .map(_.filterOrElse(_._1.hits.total == 1, ErrorMessage("Invalid nrSubmissionId"))
+          .flatMap(response => value.map((_, response._2))))
+    } else {
+      system.log.error(s"Response status ${response.status} received rom ES server for request: [$value]. Full response is: [$response]")
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala:43]: This chunk of code looks pretty similar to the equivalent chunk in `Storage`- maybe there could be a `discardBytesAndError` utility function?

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/ServiceWire.scala [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import akka.actor.typed.ActorSystem
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+import akka.stream.scaladsl.Flow
+import akka.util.ByteString
+import uk.gov.hmrc.nonrep.attachment.server.ServiceConfig
+
+import scala.concurrent.Future
+import scala.util.Try
+
+trait Request[A] {
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/service/ServiceWire.scala:13]: I like this change.

There's an argument for simplicity va a single trait to cover `Request`, `Call` and `Response` as we always use all 3 traits.

Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/ServiceWire.scala:13]: it's very unlikely to have them separated

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala [hunk 11]
```diff
@@ -24,101 +24,59 @@ import uk.gov.hmrc.nonrep.attachment.utils.JsonFormats._
...
+          val body =
+            s"""{"query": {"bool":{"must":[{"match":{"attachmentIds.keyword":"${attachmentRequestKey.request.attachmentId}"}},{"ids":{"values":"${attachmentRequestKey.request.nrSubmissionId}"}}]}}}"""
+          val request = createSignedRequest(HttpMethods.POST, config.elasticSearchUri, path, body)
+
+          system.log.info(
+            s"query for attachmentRequestKey: [$attachmentRequestKey] and notableEvents: [$notableEvents] produced path: [$path], body: [$body] and request: [$request]")
+
+          request
+      }
+      .getOrElse(HttpRequest())
```
[the remark this answers is not in the document]
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala:73]: :+1:

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala (hunk 9 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala:43]: could be turned into an utility

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala [hunk 12]
```diff
@@ -85,47 +94,114 @@ class AttachmentFlow()(implicit val system: ActorSystem[_], config: ServiceConfi
...
+        .filterOrElse(
+          entry => {
+            val (attachment, file) = entry
+            attachment.request.payloadSha256Checksum == file.toArray[Byte].calculateSha256
+          }, {
+            val error419 = StatusCodes.custom(419, "Checksum Failed")
+            ErrorMessage(error419.reason(), error419)
+          }
+        )
+        .map { case (attachment, file) => (attachment, file) }
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala:144]: I don't think the `map` on line 144 is needed.

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala (hunk 9 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/Indexing.scala:43]: actually after 2nd look I'm leaving it as it is, because only two lines can go to the utility 
- response.discardEntityBytes()
- Future.successful(Left(message))
which makes it not really useful

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/utils/CryptoUtils.scala [hunk 13]
```diff
@@ -1,8 +1,13 @@
 package uk.gov.hmrc.nonrep.attachment.utils
 
 import java.security.MessageDigest
+import java.util.Base64
+
+import org.apache.commons.codec.digest.DigestUtils
 
 object CryptoUtils {
   def calculateSha256(input: Array[Byte]): String =
```
Comment (PaulHodgson) [inline src/main/scala/uk/gov/hmrc/nonrep/attachment/utils/CryptoUtils.scala:9]: Thanks! Please could you make `ApiKey.hashedKey` delegate to this?

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala (hunk 12 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/stream/AttachmentFlow.scala:144]: good catch

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/utils/CryptoUtils.scala (hunk 13 above)
Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/utils/CryptoUtils.scala:9]: I was sure it's already there.

Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/utils/CryptoUtils.scala:9]: if not then I'll add it

File: src/main/scala/uk/gov/hmrc/nonrep/attachment/service/ServiceWire.scala (hunk 10 above)
Comment (PaulHodgson) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/ServiceWire.scala:13]: Up to you - its still good to go if you leave it

Comment (greg-gajda) [reply src/main/scala/uk/gov/hmrc/nonrep/attachment/service/ServiceWire.scala]: it's replaced now with just one trait

**Review by PaulHodgson — APPROVED** [review]
Looks good, thanks for taking the time to improve it.

**Comment by greg-gajda** [comment]
> Looks good, thanks for taking the time to improve it.

Thank you for reviewing it. 
There's still could be some minor changes when I finally run it on k8s with ATS,
but if there's anything significant, I'll notify you.
